cache/src/main/java/io/envoyproxy/controlplane/cache/SnapshotResources.java

```diff
@@ -38,4 +65,20 @@
    * Returns the version associated with this resources in this collection.
    */
   public abstract String version();
+
+  /**
+   * Returns the version associated with this resources in this collection, where the key is the name of the resource.
+   */
+  public String version(List<String> resourceNames) {
```

Comment:
Did you mean that I can change it from `List<String>` to `String`? If so, then the method logic would have to be pushed to `Snapshot#version(String typeUrl, List<String> resourceNames)`

Reply:
Ah yeah I see how this works now: you're passing the entire resource list, using a singleton list to identify EDS requests.

 One thing that I'm somewhat concerned about is that the way this works we'll use the resource map for EDS and with ADS when there's only one cluster: EDS always does one resource per request (iirc), while EDS over ADS does batches, so if the cluster size is one you'd be using the map version, but once it increased to two it would do the overall version. 

Maybe condition whether to use the individual version values vs the overall value on the `ads()` value on the watch instead? I think that might be more intuitive

Reply:
Is this a contract in Envoy that without ADS there is always one resourceName per EDS request?  If so then sure we can if on `ads()`, but what if contract changes? Maybe just a comment would show the intent of this code and cover edge case of multiple resource names without ADS? What do you think?